utils/datasets/cifar10_dataset.py

```diff
@@ -0,0 +1,330 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Library for loading and preprocessing CIFAR-10 training and testing data."""
+
+import collections
+from typing import Tuple
+import numpy as np
+import tensorflow as tf
+import tensorflow_federated as tff
+from tensorflow.keras import datasets, layers, models
+
+CIFAR_SHAPE = (32, 32, 3)
+TOTAL_FEATURE_SIZE = 32 * 32 * 3
+NUM_EXAMPLES_PER_CLIENT = 5000
+
+def load_cifar10_federated(
+    num_clients: int = 10, 
+    num_classes: int = 10, 
+    alpha: float = 1, 
+    train_client_batch_size: int = 20, 
+    test_client_batch_size: int = 100):
+  '''Construct a synthetic federated CIFAR-10 from its centralized version
+
+  Loads the train dataset into a non iid distribution over clients using the
+  sampling method based on LDA, taken from this paper:
+  Measuring the Effects of Non-Identical Data Distribution for
+  Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
+
+  Args:
+    num_clients: An integer specifing the total number of clients.
+    num_classes: An integer representing the number of lables in the dataset.
+    alpha: A float controling the data heterogeneity among clients.
+    train_client_batch_size: A float representing the batch size during 
+      training.
+    test_client_batch_size: A float representing the batch size during
+      test.
+
+  Returns:
+    A tuple of `tff.simulation.ClientData` representing unpreprocessed
+    train data and test data.
+  '''
+  (train_images, train_labels), (test_images, test_labels) \
+      = datasets.cifar10.load_data()
+  client_train = collections.OrderedDict()
+  client_test = collections.OrderedDict()
+
+  # The following structure will store the image index for each client
+  idx_batch = [[] for _ in range(num_clients)]
+
+  for k in range(num_classes):
+    # For each class(label) sample in a proportion for each client. 
+    # The proportion is determined by a dirichlet distribution.
+    label_k = np.where(train_labels.squeeze()==k)[0]
+    np.random.shuffle(label_k)
+    proportion = np.random.dirichlet(alpha*np.ones((num_clients,)))
+    proportion = proportion*(label_k.shape[0])
+    proportion = np.cumsum(proportion).astype(int)
+    split_labels = np.split(label_k, proportion)
+      
+    idx_batch = [idx_j + splitk.tolist() for idx_j, splitk \
+       in zip(idx_batch, split_labels)]
+
+  num_per_client_test = int((len(test_labels) \
+      // (num_clients*test_client_batch_size))*test_client_batch_size)
```

Comment:
I don't quite understand why you're using division without remainder here. I don't think it's important to make sure that clients have number of examples divisible by the batch size.